File: common/src/main/java/com/study/common/security/InternalRestClientConfig.java
package com.study.common.security;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.client.RestClient;

@Configuration
public class InternalRestClientConfig {

    @Bean
    public RestClient internalRestClient(
            InternalTokenProvider tokenProvider,
            @Value("${spring.application.name}") String serviceName) {
        return RestClient.builder()
                .requestInitializer(request ->
                        request.getHeaders().set("X-Internal-Token", tokenProvider.generateToken(serviceName)))
                .build();
    }
}

File: common/src/main/kotlin/com/study/common/security/InternalRestClientConfig.kt
package com.study.common.security

import org.springframework.beans.factory.annotation.Value
import org.springframework.cloud.client.loadbalancer.LoadBalanced
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.web.client.RestClient
import org.springframework.web.client.RestTemplate

@Configuration
class InternalRestClientConfig {

    @Bean
    @LoadBalanced
    fun loadBalancedRestTemplate(): RestTemplate = RestTemplate()

    @Bean
    fun internalRestClient(
        loadBalancedRestTemplate: RestTemplate,
        tokenProvider: InternalTokenProvider,
        @Value("\${spring.application.name}") serviceName: String
    ): RestClient = RestClient.builder(loadBalancedRestTemplate)
        .requestInitializer { request ->
            request.headers.set("X-Internal-Token", tokenProvider.generateToken(serviceName))
        }
        .build()
}
